pkg/codegen/pcl/binder.go

```diff
@@ -470,6 +498,188 @@ func getBooleanAttributeValue(attr *model.Attribute) (bool, *hcl.Diagnostic) {
 	}
 }
 
+// readParameterizationDescriptor parses a map of attributes that match contents of a parameterization block in the
+// form of:
+//
+//	parameterization {
+//	    name = <name>
+//	    version = <version>
+//	    value = <base64 encoded string>
+//	}
+func readParameterizationDescriptor(
+	packageName string,
+	attributes map[string]hclsyntax.Expression,
+) (*schema.ParameterizationDescriptor, *hcl.Diagnostic) {
+	descriptor := &schema.ParameterizationDescriptor{}
+
+	// read name, version, and value attributes
+	for key, value := range attributes {
+		switch key {
+		case "name":
+			name, err := evaluateLiteralExpr(value)
+			if err != nil {
+				return nil, errorf(value.Range(), "invalid name attribute of parameterization of %q: %v", packageName, err)
+			}
+			if name == "" {
+				return nil, errorf(value.Range(), "name attribute of parameterization of %q cannot be empty", packageName)
+			}
+			descriptor.Name = name
+		case "version":
+			version, _ := evaluateLiteralExpr(value)
+			parsedVersion, err := semver.Make(version)
+			if err != nil {
+				return nil, errorf(value.Range(),
+					"invalid version %q of parameterization for package %q: %v", version, packageName, err)
+			}
+			descriptor.Version = parsedVersion
+
+		case "value":
+			// value must be a base64 encoded string
+			base64EncodedValue, err := evaluateLiteralExpr(value)
+			if err != nil {
+				return nil, errorf(value.Range(), "invalid base64 encoded value for parameterization of %q: %v", packageName, err)
+			}
+
+			decoded, err := base64.StdEncoding.DecodeString(base64EncodedValue)
+			if err != nil {
+				return nil, errorf(value.Range(), "invalid base64 encoded value for parameterization of %q: %v", packageName, err)
+			}
+
+			descriptor.Value = decoded
+		}
+	}
+
+	return descriptor, nil
+}
+
+// ReadPackageDescriptors reads the package blocks in the given file and returns a map of package names to the package.
+// The descriptors blocks are top-level program blocks have the following format:
+//
+//	package <name> {
+//	  baseProviderName = <name>
+//	  baseProviderVersion = <version>
+//	  baseProviderDownloadUrl = <url>
+//	  parameterization {
+//	      name = <name>
+//	      version = <version>
+//	      value = <base64 encoded string>
+//	  }
+//	}
+//
+// These package descriptors allow the binder loader to load package schema information from a package source.
+func ReadPackageDescriptors(file *syntax.File) (map[string]*schema.PackageDescriptor, hcl.Diagnostics) {
```

Comment:
It's not quite as neat, but if we care about performance, we could pass the `map[string]*schema.PackageDescriptor` in as an argument here. That'd save the duplicate code and errors for already-defined packages and also mean we don't have to loop over the return value in the caller above. It's a bit more "mutation-y", which I don't love, but perhaps a worthwhile trade-off?

Reply:
I am not a big fan of mutation-y functions especially when it is really simple and and the performance impact is really insignificant. Don't suspect we are going to have 100s of package definitions in these files. Even then it would still be pretty fast